cvmfs/statistics_database.cc

```diff
@@ -0,0 +1,225 @@
+/**
+ * This file is part of the CernVM File System.
+ */
+
+#include "statistics_database.h"
+
+
+struct Stats {
+  std::string files_added;
+  std::string files_removed;
+  std::string files_changed;
+  std::string dir_added;
+  std::string dir_removed;
+  std::string dir_changed;
+  std::string bytes_added;
+  std::string bytes_removed;
+
+  explicit Stats(const perf::Statistics *statistics):
+    files_added(statistics->
+                    Lookup("Publish.n_files_added")->ToString()),
+    files_removed(statistics->
+                    Lookup("Publish.n_files_removed")->ToString()),
+    files_changed(statistics->
+                    Lookup("Publish.n_files_changed")->ToString()),
+    dir_added(statistics->
+                    Lookup("Publish.n_directories_added")->ToString()),
+    dir_removed(statistics->
+                    Lookup("Publish.n_directories_removed")->ToString()),
+    dir_changed(statistics->
+                    Lookup("Publish.n_directories_changed")->ToString()),
+    bytes_added(statistics->
+                    Lookup("Publish.sz_added_bytes")->ToString()),
+    bytes_removed(statistics->
+                    Lookup("Publish.sz_removed_bytes")->ToString()) {
+  }
+};
+
+const float    StatisticsDatabase::kLatestCompatibleSchema = 1.0f;
+float          StatisticsDatabase::kLatestSchema           = 1.0f;
+unsigned       StatisticsDatabase::kLatestSchemaRevision   =
+                                              RevisionFlags::kInitialRevision;
+unsigned int   StatisticsDatabase::instances               = 0;
+bool           StatisticsDatabase::compacting_fails        = false;
+
+
+namespace {
+
+  /**
+  * Get UTC Time.
+  *
+  * @return a timestamp in "YYYY-MM-DD HH:MM:SS" format
+  */
+std::string GetGMTimestamp() {
+  struct tm time_ptr;
+  char date_and_time[50];
+  time_t t = time(NULL);
+  gmtime_r(&t, &time_ptr);      // take UTC
+  // timestamp format
+  strftime(date_and_time, 50, "%Y-%m-%d %H:%M:%S", &time_ptr);
+  std::string timestamp(date_and_time);
+  return timestamp;
+}
+
+/**
+  * Build the insert statement.
+  *
+  * @param stats a struct with all values stored in strings
+  * @return the insert statement
+  */
+std::string PrepareStatement(const struct Stats &stats) {
```

Comment:
Can move to a private method of the `StatisticsDatabase` class

Reply:
In my opinion this can stay in an anonymous namespace. No point adding it to the API of the `StatisticsDatabase` class, even as a private method.

Reply:
Ok, no big deal.  Please feel free to proceed as you prefer.